### PR TITLE
NAUT-1310: add SBOM RPCs to StorageService

### PR DESCRIPTION
**Pull request by jnathangreeg**
Adds four SBOM RPCs to support node-agent (workload) and kubevuln (registry) writing SBOMs directly to the backend, per design `naut-1310-storage-deprecation.md` §5.2 / §6 S1:

- PutSBOM(unary)        — upload SBOM ≤ 4 MiB
- PutSBOMStream(client) — upload SBOM > 4 MiB, chunked
- GetSBOM(unary)        — probe (metadata_only=true) or fetch ≤ 4 MiB
- GetSBOMStream(server) — fetch > 4 MiB, chunked

Primary key on the wire is `(customer_guid, image_digest, syft_version)`, where `customer_guid` is supplied via the existing gRPC metadata header pattern used by `SendContainerProfile`. SBOMSource enum distinguishes workload / registry / host origin so a single Vulnerability Scanner pipeline downstream can route per-kind.

Hand-written StorageClient wrappers added in `storageclient.go`. Unary methods mirror the existing pattern; streaming methods marshal the SBOMSyft proto, chunk at 1 MiB, and reassemble server-side.

Makefile updated to also invoke `protoc-gen-go-grpc` — previously it ran only `protoc-gen-gogo`, leaving `storage_service_grpc.pb.go` stale on proto changes.

## Summary by CodeRabbit

* **New Features**
  * Streaming upload/download for SBOMs and container profiles with metadata-first responses and optional metadata-only retrieval.
  * Automatic chunking (~1 MiB) for large transfers; unary profile endpoints deprecated in favor of streaming.

* **Tests**
  * Expanded end-to-end tests for streaming uploads/downloads, multi-chunk reassembly, metadata-only/miss cases, and validation edge cases.

* **Chores**
  * Build tooling updated to install and use both message and gRPC code generators; help output refreshed.

[Review Change Stack](https://app.coderabbit.ai/change-stack/kubescape/backend/pull/50)

### CODE REVIEW

File: pkg/client/v1/storageclient.go [hunk 1]
```diff
@@ -486,3 +492,186 @@ func (c *StorageClient) ListNetworkNeighborhoods(ctx context.Context, namespace
 
 	return list, nil
 }
+
+// PutSBOM uploads an SBOM to the storage server. Use for payloads small
+// enough to fit in a single gRPC message (~4 MiB). For larger SBOMs, use
```
Comment (matthyx) [inline pkg/client/v1/storageclient.go]: why should the user know? cannot we always stream it?

File: pkg/client/v1/storageclient.go [hunk 2]
```diff
@@ -486,3 +492,186 @@ func (c *StorageClient) ListNetworkNeighborhoods(ctx context.Context, namespace
...
+func (c *StorageClient) PutSBOMStream(ctx context.Context, imageDigest, syftVersion string, source proto.SBOMSource, sbom *v1beta1.SBOMSyft) (*proto.PutSBOMResponse, error) {
+	if c.protoClient == nil {
+		return nil, fmt.Errorf("client is not connected")
+	}
+
+	if sbom == nil {
+		return nil, fmt.Errorf("sbom is nil")
+	}
+
+	payload, err := sbom.Marshal()
```
Comment (matthyx) [inline pkg/client/v1/storageclient.go]: is there a way to stream the marshalling too? (I find it not optimal to convert the full object and then split it)

File: pkg/client/v1/storageclient.go [hunk 3]
```diff
@@ -486,3 +492,186 @@ func (c *StorageClient) ListNetworkNeighborhoods(ctx context.Context, namespace
...
+			return nil, fmt.Errorf("failed to send chunk: %w", err)
+		}
+	}
+
+	return stream.CloseAndRecv()
+}
+
+// GetSBOM probes for or fetches an SBOM by (image_digest, syft_version).
+// Set metadataOnly=true to skip the blob and only check whether an SBOM
+// exists. For SBOMs too large to fit in a single gRPC message, use
```
Comment (matthyx) [inline pkg/client/v1/storageclient.go]: it's even more difficult to know the size of the receiving SBOM

File: pkg/client/v1/storageclient_test.go [hunk 4]
```diff
@@ -49,6 +54,34 @@ func (m *mockStorageServiceClient) ListNetworkNeighborhoods(ctx context.Context,
 	return &proto.ListNetworkNeighborhoodsResponse{Success: true}, nil
 }
 
+func (m *mockStorageServiceClient) PutSBOM(ctx context.Context, in *proto.PutSBOMRequest, opts ...grpc.CallOption) (*proto.PutSBOMResponse, error) {
```
Comment (matthyx) [inline pkg/client/v1/storageclient_test.go]: can you add REAL tests too that will validate the marshalling and unmarshalling (because in the past I had to fix the gRPC API because no one tried it)

**Review by matthyx — CHANGES_REQUESTED** [review]
I don't like the dual API depending on the size, can we make it simpler for the caller?

File: pkg/client/v1/storageclient.go [hunk 5]
```diff
@@ -486,3 +632,202 @@ func (c *StorageClient) ListNetworkNeighborhoods(ctx context.Context, namespace
...
+	if r == nil {
+		return nil, fmt.Errorf("payload reader is nil")
+	}
+
+	ctx = c.withMetadata(ctx)
+
+	if c.callTimeout != nil && *c.callTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, *c.callTimeout)
+		defer cancel()
```
Comment (matthyx) [inline pkg/client/v1/storageclient.go]: This default timeout is a good fit for short unary RPCs, but it is risky on a streaming upload whose duration legitimately depends on payload size / link speed. With the current default of 30s, a healthy large upload can still fail with `context deadline exceeded`, which undercuts the point of adding `PutSBOMStream` for the >4 MiB case.

`GetSBOMStream` explicitly avoids `callTimeout` below for the same reason. I think the upload path should do the same (and the same pattern currently exists in the container-profile stream helpers too).

```suggestion
// Note: we deliberately do NOT apply callTimeout here, because stream
// duration depends on payload size / network speed. Callers wanting a
// timeout should pass a ctx with their own deadline.
```

File: pkg/client/v1/storageclient_test.go [hunk 6]
```diff
@@ -615,3 +654,448 @@ func TestParseGRPCURL(t *testing.T) {
...
+// matthyx asked for: it stands up a real gRPC server over bufconn, ships
+// a non-trivial SBOMSyft via the client's PutSBOM, then pulls it back via
+// GetSBOM, and verifies semantic equality at every hop. Covers four
+// shapes:
+//   - Probe (metadata_only=true), exists path
+//   - Probe (metadata_only=true), miss path
+//   - Full fetch, payload fits in one chunk
+//   - Full fetch, payload is split across many small chunks (forces the
+//     reader to drain across multiple Recv calls)
+func TestStorageClient_SBOMRoundTrip(t *testing.T) {
```
Comment (matthyx) [inline pkg/client/v1/storageclient_test.go:883]: These round-trip tests validate the marshaling / chunking logic, but they only exercise fast bufconn paths. They won't catch the regression where the default 30s `callTimeout` clips a slow-but-healthy stream.

Can we add one delayed-stream test (for upload or download) that intentionally runs past 30s and proves the streaming methods are not bounded by the library default timeout?

File: pkg/client/v1/storageclient.go [hunk 7]
```diff
@@ -486,3 +632,201 @@ func (c *StorageClient) ListNetworkNeighborhoods(ctx context.Context, namespace
...
+	if r == nil {
+		return nil, fmt.Errorf("payload reader is nil")
+	}
+
+	ctx = c.withMetadata(ctx)
+
+// Note: we deliberately do NOT apply callTimeout here, because stream
+// duration depends on payload size / network speed. Callers wanting a
+// timeout should pass a ctx with their own deadline.
+	}
```
Comment (matthyx) [inline pkg/client/v1/storageclient.go]: This stray `}` closes `PutSBOMStream` early, so `stream, err := c.protoClient.PutSBOMStream(ctx)` is parsed outside the function body and the package no longer builds.

The current CI failure is:
`pkg/client/v1/storageclient.go:658:2: syntax error: non-declaration statement outside function body`

```suggestion
// Note: we deliberately do NOT apply callTimeout here, because stream
// duration depends on payload size / network speed. Callers wanting a
// timeout should pass a ctx with their own deadline.
```

**Review by matthyx — APPROVED** [review]
